src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/SmallFrozenDictionary.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Frozen
+{
+    /// <summary>Provides a <see cref="FrozenDictionary{TKey, TValue}"/> implementation to use with small item counts.</summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    /// <remarks>
+    /// No hashing here, just a straight-up linear scan that compares all the keys.
+    /// </remarks>
+    internal sealed class SmallFrozenDictionary<TKey, TValue> : FrozenDictionary<TKey, TValue>
+        where TKey : notnull
+    {
+        private readonly TKey[] _keys;
+        private readonly TValue[] _values;
+
+        internal SmallFrozenDictionary(Dictionary<TKey, TValue> source, IEqualityComparer<TKey> comparer)
+            : base(comparer)
+        {
+            Debug.Assert(source.Count != 0);
+
+            _keys = source.Keys.ToArray(source.Count);
+            _values = source.Values.ToArray(source.Count);
+        }
+
+        private protected override TKey[] KeysCore => _keys;
+        private protected override TValue[] ValuesCore => _values;
+        private protected override int CountCore => _keys.Length;
+        private protected sealed override Enumerator GetEnumeratorCore() => new Enumerator(_keys, _values);
+
+        private protected override ref readonly TValue GetValueRefOrNullRefCore(TKey key)
+        {
+            for (int i = 0; i < _keys.Length; i++)
+            {
+                if (Comparer.Equals(_keys[i], key))
```

Comment:
Would it be worth it if you created a specialized implementation for the case `TKey` is a struct and the comparer is the `Default`?

Reply:
@danmoseley There is no API change here, all these types are internal to the implementation of the existing frozen set and dictionary.

Reply:
Do you have a particular use case in mind where this would help performance?

Reply:
We have separate code paths for these cases elsewhere in the BCL, like in [`Dictionary`](https://github.com/dotnet/runtime/blob/200a90aae4905567e79aafe49380a899a8f2b0c7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L375-L440) and [`PriorityQueue`](https://github.com/dotnet/runtime/blob/200a90aae4905567e79aafe49380a899a8f2b0c7/src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs#L272-L293).

Reply:
Oh, sorry, I was thinking of something else.

Yes, this is already present. There are the ValueTypeDefaultComparerFrozen* types for this case.